tt_metal/impl/buffers/buffer.hpp

```diff
@@ -20,11 +26,97 @@ enum class BufferType {
+    SINGLE_BANK,
+    HEIGHT_SHARDED,
+    WIDTH_SHARDED,
+    BLOCK_SHARDED,
+};
+
+enum class ShardOrientation {
+    ROW_MAJOR,
+    COL_MAJOR,
+};
+
+
+struct ShardSpec {
+    CoreRangeSet shard_grid;
+    std::array<uint32_t, 2> shard_shape;
+    ShardOrientation shard_orientation = ShardOrientation::ROW_MAJOR;
+    bool halo = false;
+
+    ShardSpec(const CoreRangeSet & core_sets_,
+                    const std::array<uint32_t,2> & shard_shape_,
+                    const ShardOrientation & shard_orientation_ = ShardOrientation::ROW_MAJOR,
+                    const bool & halo_ = false):
+                    shard_grid(core_sets_), shard_shape(shard_shape_),
+                    shard_orientation(shard_orientation_), halo(halo_)
+                    {;}
+
+    const uint32_t num_cores() const { return this->shard_grid.num_cores(); }
+    const uint32_t numel() const { return this->shard_shape[0] * this->shard_shape[1]; }
+    tt::stl::reflection::Attributes attributes() const;
+
+};
+
+
+struct ShardSpecBuffer:  ShardSpec {
+    std::array<uint32_t, 2> page_shape;
+    std::array<uint32_t, 2 > tensor2d_size;
+    ShardSpecBuffer(const CoreRangeSet & core_sets_,
+                const std::array<uint32_t,2> & shard_shape_,
+                const ShardOrientation & shard_orientation_,
+                const bool & halo_,
+                const std::array<uint32_t, 2> & page_shape,
+                const std::array<uint32_t, 2> & tensor2d_shape
+                ): ShardSpec(core_sets_, shard_shape_, shard_orientation_, halo_)
+                {
+                    this->page_shape = page_shape;
+                    this-> tensor2d_size = tensor2d_shape;
+                }
+    ShardSpecBuffer(
+            const ShardSpec & shard_spec,
+            const std::array<uint32_t, 2> & page_shape,
+            const std::array<uint32_t, 2> & tensor2d_shape
+            ): ShardSpec(shard_spec)
+            {
+                this->page_shape = page_shape;
+                this-> tensor2d_size = tensor2d_shape;
+            }
+};
+
+
+struct BufferConfig {
+    Device *device;
+    uint64_t size;                 // Size in bytes
+    uint64_t page_size;            // Size of unit being interleaved. For non-interleaved buffers: size == page_size
+    BufferType buffer_type;
+    TensorMemoryLayout buffer_layout = TensorMemoryLayout::INTERLEAVED;
+};
+
+typedef BufferConfig InterleavedBufferConfig;
+
+// copied from above instead of using inheritance such that we can use
+// designator constructor
+struct ShardedBufferConfig {
+    Device *device;
+    uint64_t size;                 // Size in bytes
+    uint64_t page_size;            // Size of unit being interleaved. For non-interleaved buffers: size == page_size
+    BufferType buffer_type = BufferType::L1;
+    TensorMemoryLayout buffer_layout = TensorMemoryLayout::HEIGHT_SHARDED;
+    ShardSpecBuffer shard_parameters;
+};
+
+bool is_sharded(const TensorMemoryLayout & layout);
 class Buffer {
```

Comment:
Have you considered creating separate `InterleavedBuffer` and `ShardedBuffer` classes and have `Buffer` as a alias for `std::variant<InterleavedBuffer,ShardedBuffer`

The reason why I'm considering this pattern is because I see many `if (sharded) / else ` blocks and would probably be better encapsulated if sharded code lived in separate functions and object